pkg/server/preprocessor/auth.go

```diff
@@ -113,9 +114,18 @@ func (p *UserAuthenticator) Preprocess(payload *router.Payload, response *router
 	// For master access key, it is possible to impersonate any user of
 	// the caller's choosing.
 	if payload.HasMasterKey() {
-		if userID, ok := payload.Data["_user_id"].(string); ok {
-			payload.UserInfoID = userID
-			payload.Context = context.WithValue(payload.Context, router.UserIDContextKey, userID)
+		userID, ok := payload.Data["_user_id"].(string)
+		if !ok {
+			userID = "_god"
+		}
+		payload.UserInfoID = userID
+		payload.Context = context.WithValue(payload.Context, router.UserIDContextKey, userID)
+		// Try to create the implicitly user if it does not exist
+		info := skydb.UserInfo{
+			ID: userID,
+		}
+		if err := payload.DBConn.CreateUser(&info); err != nil && err != skydb.ErrUserDuplicated {
```

Comment:
I’d say it is the “inject user” preprocessor’s job to create a user, this is because:

* so the auth preprocessor does not have to depends on existence of a db connection
* the inject user preprocessor checks if there is a user already, and the user can be created if the user does not exists
* try creating user in each master key authenticated requests would generate many constraint violation at the db side and might result in unnecessary log entries
* constraint violation might have non-trivial performance penalty? (I am not sure about this one)